---
title: "Capsella local ancestry introgression and PopGen"
author: "Maya Wilson Brown"
date: "2024-03-25"
output: html_document
---

##### Setup 
```{r libraries}
library(stringr)
library(dplyr)
library(ggplot2)
library(ggnewscale)
library(scico)
```


```{r sourcing functions}
source("~/Documents/PhD/Research/capsella_introgression/capsella_introgression_func.R")
```

# Introgression in Capsella bursa-pastoris

#### Data Preparation

```{r capsella sequenced information}
# load capsella whole genome sequencing information
wgs <- read.csv("~/Documents/PhD/Research/capsella_sample_info/generated_mkwb/Capsella_vcf_metadata.txt", sep = "\t", header = T)
```

```{r defined populations}
#load C. bursa-pastoris population information
k3pops <- read.csv("~/Documents/PhD/Research/capsella_population_structure/cbp_pop_str.txt", header = T, sep = "\t")

# mutate another column that separates NY and NJ from the rest of the populations
ny_names <- wgs[which(wgs$citation == "R.Panko"),"vcf_sample_name"]

k3pops <- k3pops %>% mutate(k3pop_sm = case_when(vcf_sample_name %in% ny_names ~ "NYC",
                                                      .default = k3population))
```

# Viterbi decoding of ancestry

```{r load viterbi files}
# list all files with matching pattern from viterbi output folder
tbl <- list.files(path = ahmm_vit_path, pattern = "*.viterbi") 

vit <- list()  # Create an empty list to store the data

#loop to load sample data; each sample gets its own list element df
for (file in tbl) {
  file_path <- file.path(ahmm_vit_path, file)  # Get the full file path
  vit[[file]] <- read.delim(file_path, header = F, col.names = c("chrom", "start", "end", "V4", "V5", "anc_state"))  # Read the file and store in the list
}                                

#change list element names
names(vit) <- str_replace_all(names(vit), "[.]viterbi", "") #remove file suffix from name

# produces list where each element is a df of the results for one sample
```


```{r reformat viterbi list}
# add factors for ancestral state and add tract-length column
vit <- lapply(vit, function(x) {data.frame(x) %>% mutate(ancestry = ifelse(anc_state == "2,0",
                  "rubella", 
                  ifelse(anc_state == "0,2",
                         "bursa_pastoris",
                         "heterozygous")), #assign ancestry
                  tract_length = end-start)}
                  ) # calculate tract lengths
```

```{r}
# concatenate the data tables for all the samples making a new column identifying the sample
vit_all <- data.table::rbindlist(vit, idcol = "vcf_sample_name")
```

I will start by combining the viterbi decoding with the whole genome sequencing data

```{r}
# combine the sample info with viterbi decoding
vit_all <- left_join(vit_all, k3pops)

# order the scaffold levels
vit_all$chrom <- factor(vit_all$chrom, levels = paste0("jlSCF_", 9:16))
```

# Exploratory
The first thing I want to check is that the proportion of *C. rubella* ancestry is less than the proportion of *C. bursa-pastoris* ancestry. This is just a check to make sure I did not mix up the ancestry proportions from Ancestry HMM
```{r}
# proportion of local ancestry
print("Proportion of C. rubella ancestry by basepairs")
colSums(vit_all[vit_all$ancestry == "rubella", "tract_length"])/colSums(vit_all[, "tract_length"])

# bursa-pastoris
print("Proportion of C. bursa-pastoris ancestry by basepairs")
colSums(vit_all[vit_all$ancestry == "bursa_pastoris", "tract_length"])/colSums(vit_all[, "tract_length"])

# heterozygous
print("Proportion of heterozygous ancestry by basepairs")
colSums(vit_all[vit_all$ancestry == "heterozygous", "tract_length"])/colSums(vit_all[, "tract_length"])
```

Now, it might be interesting to look at the distributions of the tract lengths. I Expect them to be something close to a poisson distribution.
```{r}
# tract length distribution
tl_dist <- ggplot() + 
  geom_histogram(data=vit_all, aes(tract_length, fill=ancestry), bins = 40) +
  scale_fill_manual(values = anc.cols) +
  theme_bw() +
  facet_grid(chrom~ancestry, scales = "free") +
  theme(legend.position = "none")
tl_dist

```

```{r}
ggsave(paste0(temp_plotdir, "tract_len.png"), tl_dist, width = 7, height = 5)
```
Most tracts are short, and bursa-pastoris tracts are on average, longer than others. Heterozygous tracts are the shortest.

# Introgression Plots

Single chromosome plots in a loop

```{r}
for (scf in c(9:16)) {
  plt <- viterbi_columns_plot(df = vit_all, scaffold_num = scf, population = "NYC")
  filenm <- paste0("nyc_viterbi_scf", scf, ".png")
  ggsave(paste0(temp_plotdir, filenm), plt, units = "in", height = 7, width = 12)
}
```


```{r N. Eurasia single-scaffold introgression plots}
for (scf in c(9:16)) {
  plt <- viterbi_columns_plot(df = vit_all, scaffold_num = scf, population = "N_Europe")
  filenm <- paste0("neu_viterbi_scf", scf, ".png")
  ggsave(paste0(temp_plotdir, filenm), plt, units = "in", height = 7, width = 12)
}
```

```{r Mediterranean single-scaffold introgression plots}
for (scf in c(9:16)) {
  plt <- viterbi_columns_plot(df = vit_all, scaffold_num = scf, population = "MENA")
  filenm <- paste0("mena_iterbi_scf", scf, ".png")
  ggsave(paste0(temp_plotdir, filenm), plt, units = "in", height = 7, width = 12)
}
```

```{r East Asia single-scaffold introgression plots}
for (scf in c(9:16)) {
  plt <- viterbi_columns_plot(df = vit_all, scaffold_num = scf, population = "E_Asia")
  filenm <- paste0("easia_viterbi_scf", scf, ".png")
  ggsave(paste0(temp_plotdir, filenm), plt, units = "in", height = 7, width = 12)
}
```

```{r nyc_introgression_plot_allScaffolds}
nyc_allc <- viterbi_columns_plot_allChr(vit_all, population = "NYC", facet.col = "#585380", facet.txt.col = "white")

ggsave(paste0(temp_plotdir,"ny_viterbi_all4.png"),nyc_allc, units = "in", height = 6, width = 14)
```

```{r nyc_introgression_plot_allScaffolds}
neu_allc <- viterbi_columns_plot_allChr(vit_all, population = "N_Europe" , facet.col = "#7FC55F", facet.txt.col = "black")

# why is the order wrong >:(
ggsave(paste0(temp_plotdir,"neu_viterbi_all4.png"),neu_allc, units = "in", height = 6, width = 10)
```

```{r nyc_introgression_plot_allScaffolds}
mena_allc <- viterbi_columns_plot_allChr(vit_all, population = "MENA", facet.col = "#996330", facet.txt.col = "white")

# why is the order wrong >:(
ggsave(paste0(temp_plotdir,"mena_viterbi_all4.png"),mena_allc, units = "in", height = 6, width = 6.7)
```

```{r nyc_introgression_plot_allScaffolds}
easia_allc <- viterbi_columns_plot_allChr(vit_all, population = c("E_Asia"))

ggsave(paste0(temp_plotdir,"easia_viterbi_all4.png"),easia_allc, units = "in", height = 5, width = 5)
```


Did not get the output I expected in that there are some very small regions in the East Asian population that seem to have C. rubella ancestry.
```{r, eval=FALSE}
# load posterior for sample SRR6179241.sam (AQ415)
aq415 <- read.csv(paste0(ahmm_vit_path, "SRR6179241.sam.posterior"),
                  header = T, sep = "\t")

## have there always been very small regions shown as rubella in this population?
# list all files with matching pattern from viterbi output folder
tbl <- list.files(path = "~/Documents/PhD/Research/capsella_introgression/old_ahmm_output", pattern = "*.viterbi") 

vit <- list()  # Create an empty list to store the data

#loop to load sample data; each sample gets its own list element df
for (file in tbl) {
  file_path <- file.path("~/Documents/PhD/Research/capsella_introgression/old_ahmm_output", file)  # Get the full file path
  vit[[file]] <- read.delim(file_path, header = F, col.names = c("chrom", "start", "end", "V4", "V5", "anc_state"))  # Read the file and store in the list
}                                

#change list element names
names(vit) <- str_replace_all(names(vit), "[.]viterbi", "") #remove file suffix from name

# add factors for ancestral state and add tract-length column
vit <- lapply(vit, function(x) {data.frame(x) %>% mutate(ancestry = ifelse(anc_state == "2,0",
                  "rubella", 
                  ifelse(anc_state == "0,2",
                         "bursa_pastoris",
                         "heterozygous")), #assign ancestry
                  tract_length = end-start)}
                  ) # calculate tract lengths

# concatenate the data tables for all the samples making a new column identifying the sample
vit_all <- data.table::rbindlist(vit, idcol = "vcf_sample_name")

# combine the sample info with viterbi decoding
vit_all <- left_join(vit_all, k3pops)

# order the scaffold levels
vit_all$chrom <- factor(vit_all$chrom, levels = paste0("jlSCF_", 9:16))

east_asian <- vit_all[which(vit_all$k3population == "E_Asia"),]

```



###### Introgression .bed files
To use this tool, I need a bed file of *C. rubella* regions for each sample separately.

Starting with the NYC samples
```{r}
ny_rub_by_sample <- bed_by_sample(vit_all, anc = "rubella", population = "NYC", outdir = bed_files)
```

Now do the same for Mediterranean samples.
```{r mediterranenan samples, eval=FALSE}
mena_rub_by_sample <- bed_by_sample(vit_all, anc = "rubella",
                                  population = "MENA",
                                  outdir = bed_files)
```

And N. European samples
```{r write European to bed files}
# Warning: this function writes files automatically
eu_rub_by_sample <- bed_by_sample(vit_all, anc = "rubella",
                                  population = "N_Europe",
                                  outdir = bed_files)
```

#### Pi in introgressed regions
To look at pairwise diversity in introgressed regions, first, I assigned individual Capsella bursa-pastoris to different populations where individuals are a part of 3 main lineages: Asian, with no introgression; Mediterranean, with introgression; and European/NewYorker, with introgression.

```{r}
# generating PIXY population file for input

# read in individuals list
allsites_inds <- read.csv("~/Documents/PhD/Research/capsella_introgression/allSites_inds.txt",
                          sep = "\t", header = F)

# join with vcf metadata
allsites_dat <- left_join(allsites_inds, k3pops, join_by("V1"=="vcf_sample_name"))

# invidiual missing from population file is WAC5 which is mediterranean.
allsites_dat[which(allsites_dat$V1 == "SRR6382401.sam"), c("sample_name", "k3population", "k3pop_sm")] <- c("WAC5", "MENA", "MENA")

# Then, just save as a headerless file with vcf sample names and population names
write.table(allsites_dat[c(1,3)],
            file = "~/Documents/PhD/Research/capsella_introgression/pixy_pops.txt",
            col.names = F,
            row.names = F,
            quote = F,
            sep = "\t")
# Then, just save as a headerless file with vcf sample names and population names
write.table(allsites_dat[c(1,4)],
            file = "~/Documents/PhD/Research/capsella_introgression/pixy_pops_NYC_ownpop.txt",
            col.names = F,
            row.names = F,
            quote = F,
            sep = "\t")
```


First, I looked at pairwise diversity in each population using sliding windows of 10,000 base pairs across the whole genome.

*ANOTHER THOUGHT*
Another analysis that may be interesting is to look at PI (pairwise diversity) along the genome within NYC populations to identify regions with higher or lower PI than expected?

But first, I am interested in Dxy comparing NYC regions to other.

The first thing I need to do is create a mapping list for Pixy defined populations for comparison.

The main question here might be how extremely differentiated sites compare to genomic regions of interest (either introgressed or not)

Might want to compare the levels between the Northern European population and the Southern European population, compared to NYC and California when I get that data.


Using the PIXY documentation function in source file
```{r}
pixy_folder <- "~/Documents/PhD/Research/capsella_introgression/pixy/"
pixy_files <- list.files(pixy_folder, full.names = TRUE)
pixy_df <- pixy_to_long(pixy_files)
```

Just calculating genome wide nucleotide diversity (pi) in the different populations by just summing the numerator (differenced) and denominator (comparisons) and dividing
```{r load genome wide pi}
# calculated in windows of 10,000bp
pi_w10000 <- read.csv("~/Documents/PhD/Research/capsella_introgression/pixy/nyc_allSites_CBP_pi.txt",
                      sep = "\t",
                      header = T)
```


```{r calculate genome wide pi}
nyc_pi <- pi_w10000[which(pi_w10000$pop == "NYC" & pi_w10000$no_sites > 0),]
nyc_subgenome_pi <- sum(nyc_pi$count_diffs)/sum(nyc_pi$count_comparisons)

neu_pi <- pi_w10000[which(pi_w10000$pop == "N_Europe" & pi_w10000$no_sites > 0),]
neu_subgenome_pi <- sum(neu_pi$count_diffs)/sum(neu_pi$count_comparisons)

mena_pi <- pi_w10000[which(pi_w10000$pop == "MENA" & pi_w10000$no_sites > 0),]
mena_subgenome_pi <- sum(mena_pi$count_diffs)/sum(mena_pi$count_comparisons)

easia_pi <- pi_w10000[which(pi_w10000$pop == "E_Asia" & pi_w10000$no_sites > 0),]
easia_subgenome_pi <- sum(easia_pi$count_diffs)/sum(easia_pi$count_comparisons)
```


Pixy Plot
```{r}
pixy_df %>%
  filter(chromosome == 'jlSCF_10') %>%
  filter(statistic %in% c("avg_pi", "avg_dxy", "avg_wc_fst")) %>%
  mutate(chr_position = ((window_pos_1 + window_pos_2)/2)/1000000) %>%
  ggplot(aes(x = chr_position, y = value, color = statistic))+
  geom_line(linewidth = 0.25)+
  facet_grid(statistic ~ .,
             scales = "free_y", switch = "x", space = "free_x",
             labeller = labeller(statistic = pixy_labeller,
                                 value = label_value))+
  xlab("Position on Chromosome (Mb)")+
  ylab("Statistic Value")+
  theme_bw()+
  theme(panel.spacing = unit(0.1, "cm"),
        strip.background = element_blank(),
        strip.placement = "outside",
        legend.position = "none")+
  scale_x_continuous(expand = c(0, 0))+
  scale_y_continuous(expand = c(0, 0))+
  scale_color_brewer(palette = "Set1")
```
First, lets look at Pi with all three populations
```{r}
pixy_df %>%
  filter(chromosome == 'jlSCF_10') %>%
  filter(statistic %in% c("avg_pi")) %>%
  mutate(chr_position = ((window_pos_1 + window_pos_2)/2)/1000000) %>%
  ggplot(aes(x = chr_position, y = value, color = pop1))+
  geom_line(linewidth = 0.25)+
  facet_grid(pop1 ~ .,
             scales = "free_y", switch = "x", space = "free_x",
             labeller = labeller(statistic = pixy_labeller,
                                 value = label_value))+
  xlab("Position on Chromosome (Mb)")+
  ylab("Statistic Value")+
  theme_bw()+
  theme(panel.spacing = unit(0.1, "cm"),
        strip.background = element_blank(),
        strip.placement = "outside",
        legend.position = "none")+
  scale_x_continuous(expand = c(0, 0))+
  scale_y_continuous(expand = c(0, 0))+
  scale_color_brewer(palette = "Set1")
```
What I am really keen to see is differentiation between N. Eurasian compared to Mediterranean and actually, NYC compared to N Eurasian and Mediterranean.

Let's start with differentiation (Fst and Dxy) comparing N. Eurasian to Mediterranean 

```{r}
pixy_df %>%
  filter(chromosome == 'jlSCF_10') %>%
  filter(pop1 %in% 'N_Europe') %>% 
  filter(statistic %in% c("avg_dxy", "avg_wc_fst")) %>%
  mutate(chr_position = ((window_pos_1 + window_pos_2)/2)/1000000) %>%
  ggplot(aes(x = chr_position, y = value, color = statistic))+
  geom_line(linewidth = 0.25, alpha = 0.5)+
  facet_grid(pop2 ~ .,
             scales = "free_y", switch = "x", space = "free_x",
             labeller = labeller(statistic = pixy_labeller,
                                 value = label_value))+
  xlab("Position on Chromosome (Mb)")+
  ylab("Statistic Value")+
  theme_bw()+
  theme(panel.spacing = unit(0.1, "cm"),
        strip.background = element_blank(),
        strip.placement = "outside",
        legend.position = "none")+
  scale_x_continuous(expand = c(0, 0))+
  scale_y_continuous(expand = c(0, 0))+
  scale_color_brewer(palette = "Set1")
```

PXIY on the introgressed regions: NYC is its own population here
```{r}
pixy_folder <- "~/Documents/PhD/Research/capsella_introgression/pixy/"
pixy_files <- list.files(pixy_folder, pattern= "^[introgessed]", full.names = TRUE) # list only the files that start with 'introgressed'
pixy_df <- pixy_to_long(pixy_files)
```

```{r}
pixy_df %>%
  filter(chromosome == 'jlSCF_10') %>%
  filter(pop1 %in% 'NYC') %>% 
  filter(statistic %in% c("avg_dxy", "avg_wc_fst")) %>%
  mutate(chr_position = ((window_pos_1 + window_pos_2)/2)/1000000) %>%
  ggplot(aes(x = chr_position, y = value, color = statistic))+
  geom_point(linewidth = 0.25, alpha = 0.5)+
  facet_grid(pop2 ~ .,
             scales = "free_y", switch = "x", space = "free_x")+
  xlab("Position on Chromosome (Mb)")+
  ylab("Statistic Value")+
  theme_bw()+
  ggtitle("Differentiation in regions introgressed in NYC: Chr 10")
  theme(panel.spacing = unit(0.1, "cm"),
        strip.background = element_blank(),
        strip.placement = "outside",
        legend.position = "bottom")+
  scale_x_continuous(expand = c(0, 0))+
  scale_y_continuous(expand = c(0, 0))+
  scale_color_brewer(palette = "Set1")
```
I have a bed file so can I just do a PCA on the introgresed regions only (this is actually just the Lostruct analysis combined)



#### Genic Content
We may be interested in comparing the number of genes in introgressed regions vs non-introgressed regions. To do this, we can use BEDTools intersect to see how many basepairs of genes from a Genome Feature File (GFF) intersect with basepairs from introgressed or non-introgressed regions.

I will create BEDfiles of non-intorgressed regions as before.
