---
title: "Validation of Ancestry HMM using Lostruct"
author: "Maya Wilson Brown"
date: "2024-03-14"
output: html_document
---

Here, I will document a way that we validated that Ancestry HMM was pocking up on the kind of ancestry signals we expected.

We decided to use LOSTRUCT to look at a PCA at the locus level in our introgressed and non-introgressed loci of *Capsella bursa-pasoris*.

```{r load libraries}
# load libraries
library(lostruct)
library(vcfR)
library(tidyr)
library(stringr)
library(dplyr)
# for plotting
library(ggplot2)
library(scico)
library(ggnewscale)
library(ggrepel)
```

```{r sourcing functions}
source("~/Documents/PhD/Research/capsella_introgression/capsella_introgression_func.R")
```

Loading sample data

```{r capsella sequenced information}
# load capsella whole genome sequencing information
wgs <- read.csv("~/Documents/PhD/Research/capsella_sample_info/generated_mkwb/Capsella_vcf_metadata.txt", sep = "\t", header = T)
```

```{r defined populations}
#load C. bursa-pastoris population information
k3pops <- read.csv("~/Documents/PhD/Research/capsella_population_structure/cbp_pop_str.txt", header = T, sep = "\t")

# mutate another column that separates NY and NJ from the rest of the populations
ny_names <- wgs[which(wgs$citation == "R.Panko"),"vcf_sample_name"]

k3pops <- k3pops %>% mutate(k3pop_sm = case_when(vcf_sample_name %in% ny_names ~ "NYC",
                                                      .default = k3population))
```

### Introgressed region frequency
This section uses the generated .bed files from each *Capsella bursa-pastoris* individual and **multi-inter** from BEDTools

```{bash bedtools multiinter script, code = readLines("~/Documents/PhD/Research/capsella_introgression/multiinter.sh"), eval = FALSE}
```

Let's take a look at the output
```{r NYC multiiinter data}
# load multi inter output file
nyc_inter <- read.csv(paste0("~/Documents/PhD/Research/capsella_introgression/nyc_multiinter_rubella.txt"), header = T, sep = "\t")

# You can pass sample names to BEDtools but I will just fix them here
colnames(nyc_inter) <- str_remove(colnames(nyc_inter), "X") #remove directory info
colnames(nyc_inter) <- str_remove(colnames(nyc_inter), "_NYC_rubella.bed") #remove other info

# relevel factors
nyc_inter$chrom <- factor(nyc_inter$chrom, levels = paste0("jlSCF_", 9:16))
```

```{r NYC introgressed region frequency}
ggplot() + geom_segment(data = nyc_inter[,1:4], aes(colour= num, x=start, xend=end, y=chrom, yend=chrom), linewidth = 10) + scale_y_discrete(labels = paste0("chr ", 9:16)) + scale_color_scico(palette = 'acton', direction = -1) + theme_classic() + labs(title = "New York introgressed region frequency", colour = "individuals", x = NULL, y = "position")
```

```{r}
ggsave(paste0(plot_dir,"nyc_introgressed_freq.png"), units = "in", height = 6, width = 12)
```

Let's look at the distributions of individuals that share chunks for each scaffold to get a better idea of how common is common for each scaffold.
```{r distribution of shared regions in NYC}
ggplot() + geom_histogram(data = nyc_inter[,1:4], binwidth = 1, aes(x=num)) + facet_wrap(~chrom, scales = "free_y") +
  xlab("number of genotypes") + ggtitle("Genotypes sharing introgressed regions") + theme_bw()
```

Now again with the Mediterranean samples
```{r mediterranean multiinter data}
# load file
mena_inter <- read.delim("~/Documents/PhD/Research/capsella_introgression/mena_multiinter_rubella.txt")
# You can pass sample names to BEDtools but I will just fix them here
colnames(mena_inter) <- str_remove(colnames(mena_inter), "X") #remove directory info
colnames(mena_inter) <- str_remove(colnames(mena_inter), "_MENA_rubella.bed") #remove other info

# relevel factors
mena_inter$chrom <- factor(mena_inter$chrom, levels = paste0("jlSCF_", 9:16))
```

```{r Mediterranean introgressed region frequency}
# same plot as NYC but portrait orientation
# gradient plot where darker colors mean more individuals share an introgressed chunk
ggplot() + geom_segment(data = mena_inter[,1:4], aes(colour= num, x=chrom, xend=chrom, y=start, yend=end), linewidth = 10) + scale_color_continuous(high = "#330000", low = "#f3cea9") + theme_classic() + scale_x_discrete(labels = paste0("chr ", 9:16)) + labs(title = "Middle East and Mediterranean introgressed region frequency", colour = "individuals", x = NULL, y = "position")
```

```{r}
ggsave(paste0(plot_dir,"mena_introgressed_freq.png"), units = "in", height = 12, width = 6)
```
I will plot the N. European separately just for posterity but I am really more interested in the shared and private regions.

```{r N. Eurasian multiinter data}
# load file
neu_inter <- read.delim("~/Documents/PhD/Research/capsella_introgression/neu_multiinter_rubella.txt")
# You can pass sample names to BEDtools but I will just fix them here
colnames(neu_inter) <- str_remove(colnames(neu_inter), "X") #remove directory info
colnames(neu_inter) <- str_remove(colnames(neu_inter), "_N_Europe_rubella.bed") #remove other info

# relevel factors
neu_inter$chrom <- factor(neu_inter$chrom, levels = paste0("jlSCF_", 9:16))
```

```{r}
# gradient plot where darker colors mean more individuals share an introgressed chunk
ggplot() + geom_segment(data = neu_inter[,1:4], aes(colour= num, x=chrom, xend=chrom, y=start, yend=end), linewidth = 10) + scale_color_continuous(high = "#002C00", low = "#E5FFC5") + theme_classic() + scale_x_discrete(labels = paste0("chr ", 9:16)) + labs(title = "N. Eurasian introgressed region frequency", colour = "individuals", x = NULL, y = "position")
```

```{r}
ggsave(paste0(plot_dir,"neu_introgressed_freq.png"), units = "in", height = 12, width = 6)
```

As mentioned above, it is actually more intuitive to me and beneficial to have each population on their own number scale. I will achieve that by joining the dfs?

```{r}
# add population indicators to all
nyc_inter$population <- "NYC metro"
neu_inter$population <- "N. Eurasia"
mena_inter$population <- "Med"

# combine the relevant part of the dfs
c_inter <- list(nyc_inter[,c("chrom","start","end","num","population")],neu_inter[,c("chrom","start","end","num","population")],mena_inter[,c("chrom","start","end","num","population")])
```

Attempt to generate plot
```{r}
# make vector of pretty chromosome names
pretty_chrom <- paste0("Chr. ", 9:16)
names(pretty_chrom) <- paste0("jlSCF_", 9:16)
```


```{r}
# with ggnewscale and scico color pallets
ggplot(mapping = aes(x=population, ymin=start, ymax=end)) + 
  facet_wrap(~chrom, scales = "free", nrow =2, ncol = 4,
             labeller = labeller(chrom = pretty_chrom)) + 
  coord_flip() +
  theme_classic() +
  theme(axis.text.x = element_text(size = 4), legend.position = "bottom") + 
  labs(colour = "NYC Metro\nindividuals") +
  geom_linerange(data = c_inter[[1]], linewidth = 5, aes(color = num)) +
  scale_color_scico(palette = 'acton', direction = -1) +
  new_scale_color() +
  geom_linerange(data = c_inter[[2]], linewidth = 5, aes(color = num)) +
  scale_color_gradientn(colors = greens) +
  labs(colour = "N. Eurasia\nindividuals") +
  new_scale_color() + 
  labs(colour = "Mediterranean\nindividuals") +
  geom_linerange(data = c_inter[[3]], linewidth = 5, aes(color = num)) +
  scale_color_gradientn(colors = light.brown)
```

```{r}
ggsave(paste0(plot_dir, "introgression_plot_allscf.png"), width = 11, height = 5.5)
```


#### Evaluating introgressed regions with LOSTRUCT
The new goal here, is do have the LOSTRUCT windows to be based on genomic regions determined by introgression or not.

You can use BCF tools to pull out windows useful with lostruct.
```{bash, code = readLines("~/Documents/PhD/Research/capsella_introgression/bed_extract_multiinter.sh"), eval = FALSE}
```

Load genotype matrix for only rubella regions
```{r}
# load numeric genotype matrix from bed_extract_multiinter.sh
nyc_rubella <- read.csv("~/Desktop/nyc_rubella_vcf.txt",
                        sep = "\t",
                        header = F)

# load sample names; sample names aquired from bcftools query -l without Neslia paniculata sample
samples <- scan("~/Desktop/nyc_rubella_vcf_samples.txt", character(), quote = "")

# assign sample names to columns
colnames(nyc_rubella) <- c("chrom", "pos", samples)

# assign variant ids to rownames
rownames(nyc_rubella) <- paste(nyc_rubella$chrom, nyc_rubella$pos, sep = ":")

# split by chromosome
nyc_rubella_chr <- split(nyc_rubella[3:ncol(nyc_rubella)], nyc_rubella[1])
# split by chromosome, but keep the names
nyc_rubella_chr_snps <- split(nyc_rubella[1:2], nyc_rubella[1])

# make all data tables matrices
genomat <- lapply(nyc_rubella_chr, data.matrix)
```

I can either pass the numeric matrix directly to eigen_windows or write a window extraction function that returns the numeric matrix of a specific window. I will try to write a window extractor function.

The arguments of the function have to take a numeric integer, n, to select the window of interest. So, basically, I need to figure out the starting and stopping SNPs for

My first idea is to assign a window ID to any location where the last bp of the previous row in the multiinter BED is not the same at the first bp of the next row, indicating a gap.

Just work with Scaffold 10 for the time being.
```{r, eval = F}
# scaffold 10 matrix
scf10m <- genomat[[1]]

# grab rownames and drop the scaffold indicator
rnames <- as.integer(str_remove_all(rownames(scf10m), paste0("jlSCF_", "10", ":")))

gaps <- lead(rnames, n=1) - lag(rnames, n=1)

# find positions where theres a large gap between SNPs (10,000bp)
rnames[which(gaps > 10000)-1]
```
Kind of silly to write my own window extractor function when all of the SNPs in the matrix fall within regions that have at least one individual. So at this point, I will just make windows of about 100 SNPs. Then, for a comparison of regions without introgression, I will use BED files from the *C. bursa-pastoris* inferred regions.

```{r calculate eigens for windows on scaffold 10}
# ON scaffold 10, extract windows of 100 continuous SNPs (rows of matrix)
s10_500snp <- lostruct::eigen_windows(data = genomat[["jlSCF_10"]],
              win = 500,
              k=3)
# drop rows that are NA here as they correspond to windows without values
idx <- apply(s10_500snp, 1, function(x) all(is.na(x))) #returns T for completely empty rows
s10_500snp <- s10_500snp[!idx,] # actually drops those rows

dropped.windows <- which(idx)
```

```{r}
# go back to the names list and assign them windows?
winds <- rep(1:5000, each = 500, length.out = nrow(genomat[["jlSCF_10"]])) #the end of rep is not important and should just be a sufficiently large number; each is the number passed to win in eigen_windows, and length.out should be the number of rows in the matrix

# combine with SNP location data
windows <- cbind(nyc_rubella_chr_snps[["jlSCF_10"]], window = as.character(winds))
```

```{r reformat eigeninfo for scaffold 10}
# apply reformatting function to each row of output matrix
pca10 <- apply(s10_500snp, 1, reform_map2, dat = s10_500snp)

#rename the list elements by their window
names(pca10) <- as.character(which(idx == F))

# unroll list into long data frame
pca10_long <- bind_rows(pca10, .id = "window")

# if no population information present, fill population columns with species
pca10_long <- pca10_long %>% mutate(across(c(k3population, k3pop_sm), ~ ifelse(is.na(.x), species, .x)))

# make the windows a factor
pca10_long$window <- factor(pca10_long$window, levels = as.character(which(idx == F)))
```

# Plots
```{r}
pc_plot1 <- ggplot() +
  geom_point(data=subset(pca10_long, k3pop_sm != "Capsella bursa-pastoris" & window %in% names(pca10)[1:20]), aes(x=PC1, y=PC2, color = k3pop_sm, shape= species), size = 3, alpha = 0.5) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(values = species.cols)

pc_plot1
```


```{r}
ggsave(paste0(plot_dir, "lostruct_pca_scf10_w1-20_noempty.png"), pc_plot1, width = 10, height = 6, units = "in")
```

Single window at a time blown up
```{r}
sc10w218 <- ggplot() +
  geom_point(data=subset(pca10_long, window == "71" & k3pop_sm != "Capsella bursa-pastoris"), aes(x=PC1, y=PC2, color = k3pop_sm, shape= species), size = 3, alpha = 0.7) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(name = "population", values = species.cols, labels = pop.labels) +
  theme(strip.background =element_rect(fill="lightgrey"),
        strip.text = element_text(size = 15))
sc10w218
```

```{r}
ggsave(paste0(plot_dir,"lostruct_pca_scf10_window218.png"), sc10w218, width = 7, height = 5, units = "in")
```

```{r}
sc10w218 <- ggplot() +
  geom_point(data=subset(pca10_long, window == "71" & k3pop_sm != "Capsella bursa-pastoris"), aes(x=PC1, y=PC3, color = k3pop_sm, shape= species), size = 3, alpha = 0.7) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(name = "population", values = species.cols, labels = pop.labels) +
  theme(strip.background =element_rect(fill="lightgrey"),
        strip.text = element_text(size = 15))
sc10w218
```

To link these plots to the introgression results, now we combine the introgression plots above with shaded windows for the start and end SNPs

```{r}
sc10w10 <- ggplot() +
  geom_point(data=subset(pca10_long, window == 10 & k3pop_sm != "Capsella bursa-pastoris"), aes(x=PC1, y=PC2, color = k3pop_sm, shape= species), size = 3, alpha = 0.7) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(name = "population", values = species.cols, labels = pop.labels) +
  theme(strip.background =element_rect(fill="lightgrey"),
        strip.text = element_text(size = 15))

sc10w10
```

```{r}
ggsave("lostruct_pca_scf10_window10.png", sc10w10, width = 7, height = 5, units = "in")
```

with PC3 
```{r}
sc10w10 <- ggplot() +
  geom_point(data=subset(pca10_long, window == 10 & k3pop_sm != "Capsella bursa-pastoris"), aes(x=PC1, y=PC3, color = k3pop_sm, shape= species), size = 3, alpha = 0.7) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(name = "population", values = species.cols, labels = pop.labels) +
  theme(strip.background =element_rect(fill="lightgrey"),
        strip.text = element_text(size = 15))

sc10w10
```

```{r}
sc10w12 <- ggplot() +
  geom_point(data=subset(pca10_long, window == 12 & k3pop_sm != "Capsella bursa-pastoris"), aes(x=PC1, y=PC2, color = k3pop_sm, shape= species), size = 3, alpha = 0.7) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(name = "population", values = species.cols, labels = pop.labels) +
  theme(strip.background =element_rect(fill="lightgrey"),
        strip.text = element_text(size = 15))
sc10w12
```

```{r}
ggsave("lostruct_pca_scf10_window12.png", sc10w12, width = 10, height = 6, units = "in")
```

### The rest of LOSTRUCT
The rest of lostruct essentially makes a PCA of the local PCAs. I might be nice to be able to see what that is like.

The next steps are to compute the distance matrix on the output of eigen_windows() and then visualize it.

```{r}
# Originally computed two PCs, I think I want at least 3
# s10_500snp <- lostruct::eigen_windows(data = genomat[["jlSCF_10"]],
#               win = 500,
#               k=2)
s10_500snp <- lostruct::eigen_windows(data = genomat[["jlSCF_10"]],
              win = 500,
              k=3)

# drop rows that are NA here as they correspond to windows without values
idx <- apply(s10_500snpk3, 1, function(x) all(is.na(x))) #returns T for completely empty rows
s10_500snpk3 <- s10_500snpk3[!idx,] # actually drops those rows

# compute pairwise distance matrix between windows
pcdist <- pc_dist(s10_500snp,npc=3)
```
Cool, now just to plot them.

```{r}
fit2d <- cmdscale(pcdist, eig=TRUE, k=3 )
colnames(fit2d$points) <- c("MD1","MD2","MD3")

#plot( fit2d$points, xlab="Coordinate 1", ylab="Coordinate 2", col=rainbow(1.2*nrow(pcdist)) )
ggplot() +
  geom_point(data = fit2d$points, aes(x=MD1, y = MD2, colour = row(fit2d$points)[,1]),
             size = 3) +
  geom_text_repel(data = fit2d$points, aes(x=MD1, y = MD2, label = row(fit2d$points)[,1]), max.overlaps = 50, size = 2) +
  theme_classic()
```

```{r}
ggplot() +
  geom_point(data = fit2d$points, aes(x=MD1, y = MD3, colour = row(fit2d$points)[,1]),
             size = 3) +
  geom_text_repel(data = fit2d$points, aes(x=MD1, y = MD3, label = row(fit2d$points)[,1]), max.overlaps = 50, size = 2) +
  theme_classic()
```
For some reason, my brain needs the 3D plot right now
```{r}
library(plotly)
#make data frame
dffit <- data.frame(fit2d$points)

plot_ly(data = dffit, x=dffit$MD1, y = dffit$MD2, z = dffit$MD3, type="scatter3d", mode="markers", color = as.numeric(rownames(dffit)))
```

51,11,and 10 cluster closely for all 3 MDs.
```{r}
# which windows are they?
names(pca10)[51]
names(pca10)[11]
names(pca10)[10]
```
Let's plot them
```{r}
pc_plot1 <- ggplot() +
  geom_point(data=subset(pca10_long, k3pop_sm != "Capsella bursa-pastoris" & window %in% names(pca10)[c(51,11,10)]), aes(x=PC1, y=PC2, color = k3pop_sm, shape= species), size = 3, alpha = 0.5) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(values = species.cols)

pc_plot1
```

Looking at extreme differences
```{r}
# which windows are they?
names(pca10)[4]
names(pca10)[23]
names(pca10)[44]
```
```{r}
pc_plot1 <- ggplot() +
  geom_point(data=subset(pca10_long, k3pop_sm != "Capsella bursa-pastoris" & window %in% names(pca10)[c(4,23,44)]), aes(x=PC1, y=PC2, color = k3pop_sm, shape= species), size = 3, alpha = 0.5) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(values = species.cols)

pc_plot1
```


### Joining plots

Attempt to indicate windows corresponding to Lostruct analysis
```{r}
# only scaffold 9 for now
c_inter[[4]] <- c_inter[[1]] %>% filter(chrom == "jlSCF_9")
c_inter[[5]] <- c_inter[[2]] %>% filter(chrom == "jlSCF_9")
c_inter[[6]] <- c_inter[[3]] %>% filter(chrom == "jlSCF_9")

# only scaffold 10 for now
c_inter[[7]] <- c_inter[[1]] %>% filter(chrom == "jlSCF_10")
c_inter[[8]] <- c_inter[[2]] %>% filter(chrom == "jlSCF_10")
c_inter[[9]] <- c_inter[[3]] %>% filter(chrom == "jlSCF_10")

# only scaffold 14
inter14 <- list()
inter14[[1]] <- c_inter[[1]] %>% filter(chrom == "jlSCF_14")
inter14[[2]] <- c_inter[[2]] %>% filter(chrom == "jlSCF_14")
inter14[[3]] <- c_inter[[3]] %>% filter(chrom == "jlSCF_14")
```

The way the Lostruct eigen windows function works is that the win argument considers a window to be consecutive rows of the given matrix. So it is calculating eigenvectors at each consecutive 500 SNPs

I want to align those SNP positions with BP positions for the plot above

```{r}
# summarize the SNP table by window, just to get a min and max value
sc10_breaks <- windows %>% 
  summarise(start = min(pos),
            end = max(pos),
            .by = window)
```

```{r}
# plotting all of the windows created by LOSTRUCT
base_med <- ggplot() + geom_rect(data = sc10_breaks,
            aes(ymin = start,
                ymax = end,
                xmin = - Inf,
                xmax = Inf),
            alpha = 0.5,
            fill = "#a9ba9d") + 
  coord_flip() + 
  theme_classic() +
  geom_linerange(data = c_inter[[7]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_scico(palette = 'lajolla', direction = -1)
```

```{r}
# Plotting only the windows selected as not empty
base_med <- ggplot() + geom_rect(data = subset(sc10_breaks,
                                               window %in% names(pca10)),
            aes(ymin = start,
                ymax = end,
                xmin = - Inf,
                xmax = Inf),
            alpha = 0.5,
            fill = "#a9ba9d") + 
  coord_flip() + 
  theme_classic() +
  geom_linerange(data = c_inter[[7]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_scico(palette = 'lajolla', direction = -1)
```

Attempting with all populations. Windows are a light grey background here
```{r}
# Scaffold 10 windows for examples
ggplot() + 
  coord_flip() + 
  theme_classic() +
  ggtitle("Chr. 10 Windows of 500 SNPs") +
  theme(axis.text.x = element_text(size = 4), legend.position = "none") +
  geom_rect(data = subset(sc10_breaks, window %in% names(pca10)),
            aes(ymin = start,
                ymax = end,
                xmin = - Inf,
                xmax = Inf),
            alpha = 0.5,
            fill = "lightgrey") +
  geom_linerange(data = c_inter[[7]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_scico(palette = 'acton', direction = -1) +
  new_scale_color() + 
  geom_linerange(data = c_inter[[8]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_gradientn(colors = greens) +
  new_scale_color() + 
  geom_linerange(data = c_inter[[9]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_gradientn(colors = light.brown)
```

```{r}
ggsave(paste0(plot_dir,"scaffold10_windows500snps.png"), width = 5, height = 2)
```


We can Zoom in on a single window and check our expectations with the local PCA
```{r}
# Scaffold 10 window where NYC is not introgressed
ggplot() + 
  coord_flip() + 
  theme_classic() +
  ggtitle("Chr. 10 Windows of 500 SNPs") +
  theme(axis.text.x = element_text(size = 4), legend.position = "none") +
  geom_rect(data = subset(sc10_breaks, window %in% names(pca10)[103]),
            aes(ymin = start,
                ymax = end,
                xmin = - Inf,
                xmax = Inf),
            alpha = 0.5,
            fill = "lightgrey") +
  geom_linerange(data = c_inter[[7]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_scico(palette = 'acton', direction = -1) +
  new_scale_color() + 
  geom_linerange(data = c_inter[[8]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_gradientn(colors = greens) +
  new_scale_color() + 
  geom_linerange(data = c_inter[[9]], linewidth = 5, aes(x=population, ymin=start, ymax=end, color = num)) +
  scale_color_gradientn(colors = light.brown)
```
And to look at the corresponding PCA
```{r}
sc10w12 <- ggplot() +
  geom_point(data=subset(pca10_long, window %in% names(pca10)[103] & k3pop_sm != "Capsella bursa-pastoris"), aes(x=PC1, y=PC2, color = k3pop_sm, shape= species), size = 3, alpha = 0.7) +
  scale_shape_manual(values = c('Capsella\ grandiflora'=15, 'Capsella\ orientalis'=18, 'Capsella\ bursa-pastoris'=16, 'Capsella\ rubella'=17)) + 
  theme_classic() + facet_wrap(~window) + scale_color_manual(name = "population", values = species.cols, labels = pop.labels) +
  theme(strip.background =element_rect(fill="lightgrey"),
        strip.text = element_text(size = 15))
sc10w12
```


```{r}
# first, regions where any element of nyc is present and elements of mena and europe are not
nyc_status <- apply(all_inter["list"], 1, function(x){
# for each row, take the value of column 'list' and make it a vector and split elements into characters
el <- str_split_1(x, ",")
# make the character vector numeric
nel <- as.numeric(el)
# check if the region is private to a group
ifelse(sum(intersect(nel, nyc_indx)) >= 1, "in NYC", NA)
})
```
There is another group which I may be interested in, which is the NJ population by itself. That's population 24 and I can print the names to know they correspond to indices 43,44,45
```{r}
nj_status <- apply(all_inter["list"], 1, function(x){
# for each row, take the value of column 'list' and make it a vector and split elements into characters
el <- str_split_1(x, ",")
# make the character vector numeric
nel <- as.numeric(el)
# check if the region is private to a group
ifelse(sum(intersect(nel, 43:45)) >= 1, "in NJ", NA)
})
```

There is a more elegant way to do this (probably with casewhen) but I can fix that later. The plan is to get a vector of region status for each population
```{r}
# middle east and north africa
mena_status <- apply(all_inter["list"], 1, function(x){
# for each row, take the value of column 'list' and make it a vector and split elements into characters
el <- str_split_1(x, ",")
# make the character vector numeric
nel <- as.numeric(el)
# check if the region is private to a group
ifelse(sum(intersect(nel, mena_indx)) >= 1, "in MENA", NA)
})

# Europe
neu_status <- apply(all_inter["list"], 1, function(x){
# for each row, take the value of column 'list' and make it a vector and split elements into characters
el <- str_split_1(x, ",")
# make the character vector numeric
nel <- as.numeric(el)
# check if the region is private to a group
ifelse(sum(intersect(nel, neu_indx)) >= 1, "in N. Europe", NA)
})

# combine them
chunk_status <- data.frame(nyc_status,nj_status,mena_status,neu_status)

ind <- rowSums(is.na(chunk_status)) == ncol(chunk_status)
ind2 <- rowSums(is.na(all_stat[,6:8])) == ncol(all_stat[,6:8])
```

Bind with actual bed data and check which regions are in or out of each population
```{r}
all_stat <- cbind(all_inter[,1:5], chunk_status )
```

```{r}
# filter keeps rows that match a condition
# keep rows where europe and mena are NA (implying nyc is not, because I do not have empty regions)
seg_nyc_only <- all_stat %>%
   filter(if_all(c(mena_status,neu_status), is.na))

# present in NY/NJ and Europe; absent from MENA
seg_nyc_eu <- all_stat %>%
   filter(!(if_any(c(nyc_status, neu_status), is.na)) & is.na(mena_status))

# present in NY/NJ and MENA; absent from Europe
seg_nyc_mena <- all_stat %>%
   filter(!(if_any(c(nyc_status, mena_status), is.na)) & is.na(neu_status))

# present in NJ and Europe, which may drive the relationship between the two
seg_nj_neu <- all_stat %>%
   filter(!(if_any(c(nj_status, neu_status), is.na)) & is.na(mena_status))

# only in NJ (of which NY is inclusive)
#seg_nj <- all_stat %>%
   #filter(!is.na(nj_status) & if_all(c(neu_status, mena_status), is.na))

# private to NJ (not found elsewhere in NY)
seg_nj[which(seg_nj$num <= 3),]
```
There are `r nrow(seg_nyc_only)` introgressed regions only found in NYC, that have not been sampled in the MENA nor the N. European population.

There are introgressed regions that are only found in NJ and a sample in the Mediterranean population (namely KYRG-3-14, which is index 97 in this data): number = `r nrow(which(seg_nyc_mena$num ==4 & seg_nyc_mena$list == c(43,44,45,97)`

Visual inspection of the data shows there are only two regions that are in the N. European group and NJ and not in the rest of NYC, and those are two regions on scaffold 15
```{r}
seg_nj_neu[107:108,]

seg_nj_neu[107:108,"end"] - seg_nj_neu[107:108,"start"]
```
4k and 12k bp respectively.

Let's plot NYC only regions:
```{r}
nyc_private_hist <- ggplot() + geom_histogram(data = seg_nyc_only, aes(num)) +
  theme_classic() +
  xlab("individuals")

ggsave("~/Desktop/nyc_private_hist.png",nyc_private_hist, width = 5, height = 2)
```


```{r}
nyc_private_vio <- ggplot() + geom_boxplot(data = seg_nyc_only, aes(bp)) +
  theme_classic() +
  xlab("length (base pairs)")

ggsave("~/Desktop/nyc_private_boxplot.png",nyc_private_vio, width = 5, height = 2)
```
